msteams-platform/bots/how-to/conversations/send-proactive-messages.md

```diff
@@ -7,6 +7,153 @@ ms.author: anclear
+
+    var proactiveMessage = MessageFactory.Text($"Hello {turnContext.Activity.From.Name}. You sent me a message. This is a proactive responsive message.");
+
+    proactiveMessage.From = turnContext.Activity.From;
+
+    proactiveMessage.Conversation = new ConversationAccount
+    {
+        Id = converationReference.Id.ToString(),
+    };
+
+    await connector.Conversations.SendToConversationAsync(proactiveMessage, cancellationToken);
+
+    // Or you can use the adapter to send a message if you already have a conversation reference. You can put this code into the controller if
+    // you already have a store of conversation references. 
+    await turnContext.Adapter.ContinueConversationAsync(_appId, turnContext.Activity.GetConversationReference(), BotOnTurn, cancellationToken);
+}
+
+```
+
+You can use the adapter to send a proactive message if you already have a conversation reference; put this code into the controller:
+
+```cs
+await turnContext.Adapter.ContinueConversationAsync(_appId, turnContext.Activity.GetConversationReference(), BotOnTurn, cancellationToken);
+```
+
+To send a proactive message, the adapter requires an app ID for the bot. In a production environment, you can use the bot's app ID. In a local test environment, you can use any GUID. If the bot is not currently assigned an app ID, the notify controller self-generates a placeholder ID to use for the call.
+
+## Additional information
+
+### Avoiding 401 "Unauthorized" Errors
+
+By default, the BotBuilder SDK adds a `serviceUrl` to the list of trusted host names if the incoming request is authenticated. They are maintained in an in-memory cache. If your bot is restarted, a user awaiting a proactive message cannot receive it unless they have messaged the bot again after it restarted. 
+
+To avoid this, you must manually add the `serviceUrl` to the list of trusted host names by using: 
+
+```csharp 
+MicrosoftAppCredentials.TrustServiceUrl(serviceUrl); 
+``` 
+
+For proactive messaging, `serviceUrl` is the URL of the channel that the recipient of the proactive message is using and can be found in `Activity.ServiceUrl`. 
+You'll want to add the above code just prior to the the code that sends the proactive message. In the above example:
+
+```cs
+var proactiveMessage = MessageFactory.Text($"Hello {turnContext.Activity.From.Name}. You sent me a message. This is a proactive responsive message.");
+```
+
+For more information and a related example code, see Bot Framework SDK [Send proactive notifications to users](https://docs.microsoft.com/azure/bot-service/bot-builder-howto-proactive-message?view=azure-bot-service-4.0&tabs=csharp) article.
+
+
+## Additional resources
+
+- [Send proactive notifications to users](https://docs.microsoft.com/azure/bot-service/bot-builder-howto-proactive-message?view=azure-bot-service-4.0&tabs=csharp) - Bot Framework SDK proactive messages explained
```

Comment:
Question: What would you need to do to write a bot that works well with proactive messages on **Teams** plus other channels?